composer/utils/remote_uploader.py

```diff
@@ -0,0 +1,172 @@
+from typing import List
+
+from composer.utils.dist import get_local_rank
+from composer.utils.file_helpers import (
+    maybe_create_object_store_from_uri,
+)
+from composer.utils.object_store.object_store import ObjectStoreTransientError
+from composer.utils.retrying import retry
+
+log = logging.getLogger(__name__)
+
+__all__ = ['RemoteUploader']
+
+
+def _upload_worker(
+    remote_folder: str,
+    file_queue: queue.Queue,
+    is_closed_event: multiprocessing.Event,  # pyright: ignore[reportGeneralTypeIssues]
+    num_attempts: int,
+    parent_process_id: int,
+) -> None:
+    object_store = maybe_create_object_store_from_uri(remote_folder)
+    if object_store is None:
+        raise RuntimeError(f'Can not create object store from remote folder {remote_folder}')
+
+    @retry(ObjectStoreTransientError, num_attempts=num_attempts)
+    def upload_file(
+        remote_file_name: str,
+        local_file_path: str,
+        overwrite: bool,
+        retry_index: int = 0,
+    ):
+        if retry_index == 0 and not overwrite:
+            try:
+                object_store.get_object_size(remote_file_name)
+            except FileNotFoundError:
+                # Good! It shouldn't exist.
+                pass
+            else:
+                raise FileExistsError(f'Object {remote_file_name} already exists, but overwrite was set to False.')
+        log.info(f'Uploading file {local_file_path} to {remote_file_name}')
+        object_store.upload_object(
+            object_name=remote_file_name,
+            filename=local_file_path,
+        )
+        os.remove(local_file_path)
+
+    def is_parent_process_dead():
+        try:
+            # Sending signal 0 to ppid will raise OSError if ppid is not running
+            # and do nothing otherwise
+            os.kill(parent_process_id, 0)
+        except OSError:
+            return True
+        return False
+
+    while True:
+        try:
+            file_path_to_upload, remote_file_name, overwrite = file_queue.get(block=True, timeout=0.5)
+        except queue.Empty:
+            if is_closed_event.is_set():
+                break
+            elif is_parent_process_dead():
+                log.warning(f'Terminate the uploader worker since parent process died')
+                break
+            else:
+                continue
+
+        # When encountering issues with too much concurrency in uploads, staggering the uploads can help.
+        # This stagger is intended for use when uploading model shards from every rank, and will effectively reduce
+        # the concurrency by a factor of num GPUs per node.
+        local_rank = get_local_rank()
+        local_rank_stagger = int(os.environ.get('COMPOSER_LOCAL_RANK_STAGGER_SECONDS', 0))
+        log.debug(f'Staggering uploads by {local_rank * local_rank_stagger} seconds on {local_rank} local rank.')
+        time.sleep(local_rank * local_rank_stagger)
+
+        upload_file(remote_file_name, file_path_to_upload, overwrite)
+
```

Comment:
we also want to the worker to upload the symlink if specified

Reply:
The remote uploader is the basic class to upload any file. The symlink file upload is in the checkpoint_saver in this PR, https://github.com/mosaicml/composer/pull/3320